### PR TITLE
Make `bark` could have tiny model

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Make `bark` could have tiny model. This is mainly for #24952

cc @ylacombe

### CODE REVIEW

File: tests/models/bark/test_modeling_bark.py [hunk 1]
```diff
@@ -137,14 +135,15 @@ def get_config(self):
     def get_pipeline_config(self):
         config = self.get_config()
         config.vocab_size = 300
+        config.output_vocab_size = 300
```
Comment (ydshieh) [inline tests/models/bark/test_modeling_bark.py:138]: I don't know why, but if they are different values, there is an issue when trying to load a model after saving it. Probably something to look in the future (maybe just a false alarm)

File: src/transformers/models/bark/configuration_bark.py [hunk 2]
```diff
@@ -301,7 +301,8 @@ def __init__(
         self.semantic_config = BarkSemanticConfig(**semantic_config)
         self.coarse_acoustics_config = BarkCoarseConfig(**coarse_acoustics_config)
         self.fine_acoustics_config = BarkFineConfig(**fine_acoustics_config)
-        self.codec_config = AutoConfig.for_model(**codec_config)
+        codec_model_type = codec_config["model_type"] if "model_type" in codec_config else "encodec"
+        self.codec_config = CONFIG_MAPPING[codec_model_type](**codec_config)
```
Comment (ydshieh) [inline src/transformers/models/bark/configuration_bark.py:305]: We need to provide a default model type. This is similar to `Blip2Config`.

`AutoConfig.for_model(**codec_config)` will fail if no `model_type` is in it.

File: src/transformers/models/bark/configuration_bark.py [hunk 3]
```diff
@@ -313,7 +314,7 @@ def from_sub_model_configs(
         semantic_config: BarkSemanticConfig,
         coarse_acoustics_config: BarkCoarseConfig,
         fine_acoustics_config: BarkFineConfig,
-        codec_config: AutoConfig,
+        codec_config: PretrainedConfig,
```
Comment (ydshieh) [inline src/transformers/models/bark/configuration_bark.py:317]: `AutoConfig` is not the type we expose to be used for models.

File: tests/models/bark/test_modeling_bark.py [hunk 4]
```diff
@@ -72,8 +74,6 @@ def __init__(
         initializer_range=0.02,
         n_codes_total=8,  # for BarkFineModel
         n_codes_given=1,  # for BarkFineModel
-        config_class=None,
-        model_class=None,
```
Comment (ydshieh) [inline tests/models/bark/test_modeling_bark.py:76]: There is no need to have these arguments. We can just set it explicitly in each model tester classes.

File: tests/models/bark/test_modeling_bark.py [hunk 5]
```diff
@@ -121,7 +119,7 @@ def prepare_config_and_inputs(self):
         return config, inputs_dict
 
     def get_config(self):
-        return self.config_class(
+        return BarkSemanticConfig(
```
Comment (ydshieh) [inline tests/models/bark/test_modeling_bark.py:122]: just like this.

File: tests/models/bark/test_modeling_bark.py [hunk 6]
```diff
@@ -473,6 +466,79 @@ def create_and_check_decoder_model_past_large_inputs(self, config, inputs_dict):
         self.parent.assertTrue(torch.allclose(output_from_past_slice, output_from_no_past_slice, atol=1e-3))
 
 
+class BarkModelTester:
```
Comment (ydshieh) [inline tests/models/bark/test_modeling_bark.py:469]: This is necessary to be able to create tiny models.

File: utils/create_dummy_models.py [hunk 7]
```diff
@@ -974,6 +974,10 @@ def get_token_id_from_tokenizer(token_id_name, tokenizer, original_token_id):
 
 
 def get_config_overrides(config_class, processors):
+    # `Bark` configuration is too special. Let's just not handle this for now.
+    if config_class.__name__ == "BarkConfig":
+        return {}
+
```
Comment (ylacombe) [inline utils/create_dummy_models.py:980]: Out of curiosity, what is the purpose of the function `get_config_overrides` ?

**Review by ylacombe — COMMENTED** [review]
Hi @ydshieh, many thanks for your help in that matter!
Looks great to me, I will implement some actual tests ASAP

File: utils/create_dummy_models.py (hunk 7 above)
Comment (ydshieh) [reply utils/create_dummy_models.py:980]: This is tiny model creation - which somehow messy. But `get_config_overrides` here is to use the values from the tokenizer in the config, like `vocab_size`. (note, this is not real model, and during creation, we use dummy values, but still need some consistency between tokenizer (or processor) and config)

**Comment by ydshieh** [comment]
> Hi @ydshieh, many thanks for your help in that matter! Looks great to me, I will implement some actual tests ASAP

I think you already tests in your PR. I will try to merger your PR with my own, see how those tests go and let you know.

**Comment by ydshieh** [comment]
> > Hi @ydshieh, many thanks for your help in that matter! Looks great to me, I will implement some actual tests ASAP
> 
> I think you already tests in your PR. I will try to merger your PR with my own, see how those tests go and let you know.

Oh, you mean the `TODO` I put in this PR. Nice, thanks in advance

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this!

**Comment by ydshieh** [comment]
Going to merge. But `bark` is kind special and the tiny model for it still has some issue.

**Pull request merged**